### PR TITLE
Refactor IJupyterSession restarts & constructors

### PR DESCRIPTION
**Pull request by DonJayamanne**
Would like to clean up before i implement auto restarting local kernels.

### CODE REVIEW

File: src/client/datascience/baseJupyterSession.ts [hunk 1]
```diff
@@ -51,7 +51,6 @@ export abstract class BaseJupyterSession implements IJupyterSession {
     protected get session(): ISessionWithSocket | undefined {
         return this._session;
     }
-    protected kernelConnectionMetadata?: KernelConnectionMetadata;
```
Comment (DonJayamanne) [inline src/client/datascience/baseJupyterSession.ts:54]: This cannot be undefined, 
Moved into constructor

File: src/client/datascience/baseJupyterSession.ts [hunk 2]
```diff
@@ -85,10 +84,10 @@ export abstract class BaseJupyterSession implements IJupyterSession {
 
     constructor(
         protected resource: Resource,
+        protected readonly kernelConnectionMetadata: KernelConnectionMetadata,
         private restartSessionUsed: (id: Kernel.IKernelConnection) => void,
         public workingDirectory: string,
-        private readonly interruptTimeout: number,
-        private readonly restartTimeout: number
+        private readonly interruptTimeout: number
```
Comment (DonJayamanne) [inline src/client/datascience/baseJupyterSession.ts:90]: restart is handled by sub class, hence remvoed from here

File: src/client/datascience/baseJupyterSession.ts [hunk 3]
```diff
@@ -282,13 +281,7 @@ export abstract class BaseJupyterSession implements IJupyterSession {
     }
 
     // Sub classes need to implement their own restarting specific code
-    protected abstract startRestartSession(timeout: number): void;
-    // Sub classes need to implement their own kernel change specific code
-    protected abstract createNewKernelSession(
```
Comment (DonJayamanne) [inline src/client/datascience/baseJupyterSession.ts:287]: This method isn't called from sub class, only ever class in jupyter class (method was never used in raw kernels)

File: src/client/datascience/jupyter/jupyterSession.ts [hunk 4]
```diff
@@ -34,7 +34,7 @@ export class JupyterSession extends BaseJupyterSession {
     constructor(
         resource: Resource,
         private connInfo: IJupyterConnection,
-        kernelSpec: KernelConnectionMetadata | undefined,
+        kernelConnectionMetadata: KernelConnectionMetadata,
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/jupyterSession.ts:37]: Kerne spec cannot be undefined.

File: src/client/datascience/jupyter/jupyterSession.ts [hunk 5]
```diff
@@ -58,21 +56,13 @@ export class JupyterSession extends BaseJupyterSession {
         return this.waitForIdleOnSession(this.session, timeout);
     }
 
-    public async connect(timeoutMs: number, cancelToken?: CancellationToken, disableUI?: boolean): Promise<void> {
+    public async connect(cancelToken?: CancellationToken, disableUI?: boolean): Promise<void> {
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/jupyterSession.ts:59]: Timeout is passed in ctor, hence no need to pass this around.

File: src/client/datascience/jupyter/jupyterSession.ts [hunk 6]
```diff
@@ -82,40 +72,34 @@ export class JupyterSession extends BaseJupyterSession {
     }
 
     public async createNewKernelSession(
-        resource: Resource,
-        kernelConnection: KernelConnectionMetadata | undefined,
-        timeoutMS: number,
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/jupyterSession.ts:87]: All of this information is specific to this class (JupyterSession), never changing, hence passed into ctor and we have them in properties.

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/jupyter/jupyterSession.ts [hunk 7]
```diff
@@ -82,40 +72,34 @@ export class JupyterSession extends BaseJupyterSession {
...
         cancelToken?: CancellationToken,
         disableUI?: boolean
     ): Promise<ISessionWithSocket> {
         let newSession: ISessionWithSocket | undefined;
-
-        // update resource as we know it now.
-        this.resource = resource;
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/jupyterSession.ts:94]: This is now gone, this is no longer required, in the ctor we define this property, hence its never nullable.

File: src/client/datascience/raw-kernel/rawJupyterSession.ts [hunk 8]
```diff
@@ -63,33 +64,27 @@ export class RawJupyterSession extends BaseJupyterSession {
...
+            sendKernelTelemetryEvent(this.resource, Telemetry.RawKernelSessionStartSuccess);
+            sendKernelTelemetryEvent(this.resource, Telemetry.RawKernelSessionStart, stopWatch.elapsedTime);
             traceInfo('Raw session started and connected');
             this.setSession(newSession);
 
             // Listen for session status changes
             this.session?.statusChanged.connect(this.statusHandler); // NOSONAR
 
-            // Update kernelspec and interpreter
-            this.kernelConnectionMetadata = newSession.kernelProcess?.kernelConnectionMetadata;
```
Comment (DonJayamanne) [inline src/client/datascience/raw-kernel/rawJupyterSession.ts:91]: This is now gone, this is no longer required, in the ctor we define this property, hence its never nullable.
I.e. we always have the right value when the class is instantiated

File: src/client/datascience/raw-kernel/rawJupyterSession.ts [hunk 9]
```diff
@@ -154,35 +149,13 @@ export class RawJupyterSession extends BaseJupyterSession {
...
         } finally {
-            sendKernelTelemetryEvent(resource, Telemetry.RawKernelSessionConnect, stopWatch.elapsedTime);
+            sendKernelTelemetryEvent(this.resource, Telemetry.RawKernelSessionConnect, stopWatch.elapsedTime);
         }
 
         this.connected = true;
         return newSession.kernelProcess.kernelConnectionMetadata;
     }
 
-    public async createNewKernelSession(
```
Comment (DonJayamanne) [inline src/client/datascience/raw-kernel/rawJupyterSession.ts:164]: This was never used.

File: src/client/datascience/raw-kernel/rawSession.ts [hunk 10]
```diff
@@ -39,6 +39,19 @@ export class RawSession implements ISessionWithSocket {
     private readonly _connectionStatusChanged: Signal<this, Kernel.ConnectionStatus>;
     private readonly exitHandler: IDisposable;
     private readonly signaling: typeof import('@lumino/signaling');
+    private _jupyterLabServices?: typeof import('@jupyterlab/services')
+    private cellExecutedSuccessfully?: boolean;
+    public get AtleastOneCellExecutedSuccessfully() {
```
Comment (DonJayamanne) [inline src/client/datascience/raw-kernel/rawSession.ts:44]: Damn it, was meant for another PR, but leaving this.
Basically this allows me to tell whether we successfully executed at least one cell (user code) against this session/kernel.

This is required to implement the logic we discussed (restart only if we have successfully run a cell ,etc...)
that stuff is in a separate PR.

File: src/client/datascience/types.ts [hunk 11]
```diff
@@ -281,7 +281,7 @@ export interface IJupyterSessionManager extends IAsyncDisposable {
     readonly onRestartSessionUsed: Event<Kernel.IKernelConnection>;
     startNew(
         resource: Resource,
-        kernelConnection: KernelConnectionMetadata | undefined,
+        kernelConnection: KernelConnectionMetadata,
```
Comment (DonJayamanne) [inline src/client/datascience/types.ts:284]: Yay, no longer undefined.

**Pull request merged**